### PR TITLE
MAILBOX-379 PreDeletionHook API & plug into MessageManagers

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/MetadataWithMailboxId.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        this.mailboxId = mailboxId;
+    }
+
+    public MessageMetaData getMessageMetaData() {
+        return messageMetaData;
+    }
+
+    public MailboxId getMailboxId() {
+        return mailboxId;
+    }
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/MetadataWithMailboxId.java]: No equals + hashCode ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/extension/PreDeletionHook.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+import org.apache.james.mailbox.store.MetadataWithMailboxId;
+import org.reactivestreams.Publisher;
+
+public interface PreDeletionHook {
+
+    class DeletionId {
+        private final UUID id;
+
+        public DeletionId() {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/extension/PreDeletionHook.java]: This should be in a `PreDeletionHook.random()`. We should also get `from(uuid)` && `from(String)`

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/extension/PreDeletionHook.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            return false;
+        }
+
+        @Override
+        public final int hashCode() {
+            return Objects.hash(id);
+        }
+    }
+
+    class DeleteOperation {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/extension/PreDeletionHook.java]: equals & hashCode ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 4]
```diff
@@ -95,17 +98,26 @@
     private final MessageId.Factory messageIdFactory;
     private final QuotaManager quotaManager;
     private final QuotaRootResolver quotaRootResolver;
+    private final Set<PreDeletionHook> preDeletionHooks;
 
     @Inject
+    public StoreMessageIdManager(MailboxManager mailboxManager, MailboxSessionMapperFactory mailboxSessionMapperFactory,
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: Can we try keeping a single constructor in this class?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java [hunk 5]
```diff
@@ -163,6 +186,74 @@ public void deleteShouldNotCallEventDispatcherWhenMessageIsInWrongMailbox() thro
...
+            .thenThrow(new RuntimeException("throw at hook 1"));
+
+        MessageId messageId = testingData.persist(mailbox1.getMailboxId(), messageUid1, FLAGS, session);
+        assertThatThrownBy(() -> messageIdManager.delete(messageId, ImmutableList.of(mailbox1.getMailboxId()), session))
+            .isInstanceOf(RuntimeException.class);
+
+        assertThat(messageIdManager.getMessages(ImmutableList.of(messageId), FetchGroupImpl.MINIMAL, session)
+            .stream()
+            .map(MessageResult::getMessageId))
+        .hasSize(1)
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerSideEffectTest.java]: indent fail

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/extension/PreDeletionHookTest.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+        @Test
+        void shouldMatchBeanContract() {
+            EqualsVerifier.forClass(PreDeletionHook.DeletionId.class)
+                .verify();
+        }
+    }
+
+    @Nested
+    class DeleteOperation {
```
Comment (trantienduchn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/extension/PreDeletionHookTest.java]: s/DeleteOperation/DeleteOperationTest

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/MetadataWithMailboxId.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+package org.apache.james.mailbox.store;
+
+import java.util.Objects;
+
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.mailbox.model.MessageMetaData;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+
+public class MetadataWithMailboxId {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/MetadataWithMailboxId.java]: Shouldn't this be in `mailbox/api` ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/extension/PreDeletionHook.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+
+import org.apache.james.mailbox.store.MetadataWithMailboxId;
+import org.reactivestreams.Publisher;
+
+import com.google.common.collect.ImmutableSet;
+
+public interface PreDeletionHook {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/extension/PreDeletionHook.java]: Shouldn't this be in `mailbox/api` ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 9]
```diff
@@ -688,10 +697,23 @@ public MessageResultIterator getMessages(MessageRange set, FetchGroup fetchGroup
...
+                .map(mailboxMessage -> MetadataWithMailboxId.from(mailboxMessage.metaData(), mailboxMessage.getMailboxId()))
+                .collect(Guavate.toImmutableList()));
+
+        return Flux.fromIterable(preDeletionHooks)
+            .flatMap(hook -> hook.notifyDelete(deleteOperation))
+            .then(Mono.fromSupplier(Throwing.supplier(() -> deleteMarkedInMailbox(range, messageMapper))))
+            .block();
+    }
 
+    private Map<MessageUid, MessageMetaData> deleteMarkedInMailbox(MessageRange range, MessageMapper messageMapper) throws MailboxException {
```
Comment (trantienduchn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: `deleteMarkedInMailbox()` should take `messagesWillBeDeleted` parameter but it is a task from Refactoring expunge PR. Wait for the cherry-pick

File: mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/JPAMailboxManagerTest.java [hunk 10]
```diff
@@ -26,10 +26,17 @@
 import org.apache.james.mailbox.jpa.openjpa.OpenJPAMailboxManager;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 
 public class JPAMailboxManagerTest extends MailboxManagerTest<OpenJPAMailboxManager> {
 
+    @Disabled("Maildir is using DefaultMessageId which doesn't support full feature of a messageId, which is an essential" +
```
Comment (trantienduchn) [inline mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/JPAMailboxManagerTest.java]: s/Maildir/JPA mailbox manager

**Comment by trantienduchn** [comment]
```
[ec9a5420ac584db5e9840beeb69ff31a0d81e0aa] [ERROR] org.apache.james.mailbox.spring.SpringMailboxTest  Time elapsed: 3.195 s  <<< ERROR!
[ec9a5420ac584db5e9840beeb69ff31a0d81e0aa] org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'maildir-mailboxmanager' defined in class path resource [META-INF/spring/mailbox-maildir.xml]: Unsatisfied dependency expressed through constructor argument with index 11 of type [java.util.Set]: Ambiguous constructor argument types - did you specify the correct bean references as constructor arguments?
[ec9a5420ac584db5e9840beeb69ff31a0d81e0aa] 	at org.apache.james.mailbox.spring.SpringMailboxTest.beforeClass(SpringMailboxTest.java:35)
```

File: mailbox/maildir/src/main/resources/META-INF/spring/mailbox-maildir.xml [hunk 11]
```diff
@@ -1,67 +1,68 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chibenwa) [inline mailbox/maildir/src/main/resources/META-INF/spring/mailbox-maildir.xml]: Please do these line-break changes in a separate commits

File: mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/DomainUserMaildirMailboxManagerTest.java [hunk 12]
```diff
@@ -22,10 +22,18 @@
...
 import org.apache.james.mailbox.events.EventBus;
 import org.apache.james.mailbox.store.StoreMailboxManager;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
 public class DomainUserMaildirMailboxManagerTest extends MailboxManagerTest<StoreMailboxManager> {
 
+    @Disabled("Maildir is using DefaultMessageId which doesn't support full feature of a messageId, which is an essential" +
+        "element of the Vault")
```
Comment (Arsnael) [inline mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/DomainUserMaildirMailboxManagerTest.java]: missing space between the two strings concatenation

File: mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/FullUserMaildirMailboxManagerTest.java [hunk 13]
```diff
@@ -22,9 +22,18 @@
...
 import org.apache.james.mailbox.events.EventBus;
 import org.apache.james.mailbox.store.StoreMailboxManager;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
 public class FullUserMaildirMailboxManagerTest extends MailboxManagerTest<StoreMailboxManager> {
+
+    @Disabled("Maildir is using DefaultMessageId which doesn't support full feature of a messageId, which is an essential" +
+        "element of the Vault")
```
Comment (Arsnael) [inline mailbox/maildir/src/test/java/org/apache/james/mailbox/maildir/FullUserMaildirMailboxManagerTest.java]: Idem

File: mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/JPAMailboxManagerTest.java [hunk 14]
```diff
@@ -26,10 +26,17 @@
 import org.apache.james.mailbox.jpa.openjpa.OpenJPAMailboxManager;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 
 public class JPAMailboxManagerTest extends MailboxManagerTest<OpenJPAMailboxManager> {
 
+    @Disabled("JPAMailboxManager is using DefaultMessageId which doesn't support full feature of a messageId, which is an essential" +
+        "element of the Vault")
```
Comment (Arsnael) [inline mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/JPAMailboxManagerTest.java]: Idem

File: mailbox/maildir/src/main/resources/META-INF/spring/mailbox-maildir.xml [hunk 15]
```diff
@@ -1,67 +1,68 @@
...
+        <constructor-arg index="2" ref="memory-locker"/>
+        <constructor-arg index="3" ref="messageParser"/>
+        <constructor-arg index="4" ref="messageIdFactory" />
+        <constructor-arg index="5" ref="storeMailboxAnnotationManager" />
+        <constructor-arg index="6" ref="event-bus" />
+        <constructor-arg index="7" ref="storeRightManager" />
+        <constructor-arg index="8" ref="quotaComponents" />
+        <constructor-arg index="9" ref="indexer" />
+        <constructor-arg index="10" ref="no-mailbox-configuration" />
+        <constructor-arg index="11"><set/></constructor-arg>
```
Comment (Arsnael) [inline mailbox/maildir/src/main/resources/META-INF/spring/mailbox-maildir.xml]: Should'nt we need a ref like for the other arguments?

File: mailbox/memory/src/main/resources/META-INF/spring/mailbox-memory.xml [hunk 16]
```diff
@@ -1,57 +1,58 @@
...
+        <constructor-arg index="1" ref="sessionProvider"/>
+        <constructor-arg index="2" ref="memory-locker"/>
+        <constructor-arg index="3" ref="messageParser"/>
+        <constructor-arg index="4" ref="messageIdFactory" />
+        <constructor-arg index="5" ref="event-bus" />
+        <constructor-arg index="6" ref="storeMailboxAnnotationManager" />
+        <constructor-arg index="7" ref="storeRightManager" />
+        <constructor-arg index="8" ref="quotaComponents" />
+        <constructor-arg index="9" ref="indexer" />
+        <constructor-arg index="10"><set/></constructor-arg>
```
Comment (Arsnael) [inline mailbox/memory/src/main/resources/META-INF/spring/mailbox-memory.xml]: ref?

File: mailbox/maildir/src/main/resources/META-INF/spring/mailbox-maildir.xml (hunk 15 above)
Comment (chibenwa) [reply mailbox/maildir/src/main/resources/META-INF/spring/mailbox-maildir.xml]: No if it works it's cool. (its different because we inject a set @Arsnael ) Thanks @trantienduchn for searching Spring graveyard and doing findings like this one. It might solve some other similar problems as well.

File: mailbox/api/src/main/java/org/apache/james/mailbox/extension/PreDeletionHook.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+public interface PreDeletionHook {
+
+    class DeletionId {
+
+        public static DeletionId random() {
+            return new DeletionId(UUID.randomUUID());
+        }
+
+        public static DeletionId of(UUID uuid) {
+            return new DeletionId(uuid);
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/extension/PreDeletionHook.java]: Null not checked.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 18]
```diff
@@ -688,10 +697,23 @@ public MessageResultIterator getMessages(MessageRange set, FetchGroup fetchGroup
 
     }
 
-    protected Map<MessageUid, MessageMetaData> deleteMarkedInMailbox(final MessageRange range, MailboxSession session) throws MailboxException {
-
+    protected Map<MessageUid, MessageMetaData> deleteWithPreHooks(final MessageRange range, MailboxSession session) throws MailboxException {
         final MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
+        List<MailboxMessage> messagesWillBeDeleted = ImmutableList.copyOf(
+            messageMapper.findInMailbox(mailbox, range, FetchType.Metadata, -1));
```
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: Can you add the information that *-1* means no limit.
Maybe you can introduce a constant for that in that *Limit* class.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 19]
```diff
@@ -688,10 +697,23 @@ public MessageResultIterator getMessages(MessageRange set, FetchGroup fetchGroup
...
+
+        DeleteOperation deleteOperation = DeleteOperation.from(
+            messagesWillBeDeleted.stream()
+                .map(mailboxMessage -> MetadataWithMailboxId.from(mailboxMessage.metaData(), mailboxMessage.getMailboxId()))
+                .collect(Guavate.toImmutableList()));
+
+        return Flux.fromIterable(preDeletionHooks)
+            .flatMap(hook -> hook.notifyDelete(deleteOperation))
```
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: Deleting a huge range of mails on a configuration with several hooks will have an impact on deletion.

Comment (trantienduchn) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: what do you mean `impact`? are you mentioning about performance issue? I could place `.parallel()` in flux pipeline to optimize hooks execution if you agree?

Comment (aduprat) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: It's just a reminder, I don't have a solution; and we don't have a problem yet ;)

There are multiple solution for such problems, wait and see.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 18 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: We should not forget to refactor this once @Arsnael  PR is merged but OK for now.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 19 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: > Deleting a huge range of mails on a configuration with several hooks will have an impact on deletion.

This is necessary: we need hook execution to be synchronous. That's why it's hooks and not listeners.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 18 above)
Comment (trantienduchn) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: there is no dedicated `Limit` class for message search. But I found `UNLIMITED` constant one

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Comment by chibenwa** [comment]
```
[90bd3333f2724499de8e6831ebb89160ada0de5e] [ERROR]   VacationIntegrationTest.jmapVacationShouldNotSendNotificationTwice:243->assertOneMessageReceived:393 1 expectation failed.
[90bd3333f2724499de8e6831ebb89160ada0de5e] JSON path [1][1].list doesn't match.
[90bd3333f2724499de8e6831ebb89160ada0de5e] Expected: a collection with size <1>
[90bd3333f2724499de8e6831ebb89160ada0de5e]   Actual: [{mailboxIds=[5], from={name=benwa@domain.tld, email=benwa@domain.tld}, textBody=Message explaining my wonderful vacations, id=5, to=[{name=matthieu@domain.tld, email=matthieu@domain.tld}]}, {mailboxIds=[5], from={name=benwa@domain.tld, email=benwa@domain.tld}, textBody=Message explaining my wonderful vacations, id=6, to=[{name=matthieu@domain.tld, email=matthieu@domain.tld}]}]
```

Don't seem much related

test this please

**Comment by Arsnael** [comment]
Looks ok to me. I don't put approval yet to not forget that it needs to be refactored after expunge refactoring has been approved

File: mailbox/memory/src/main/resources/META-INF/spring/mailbox-memory.xml (hunk 16 above)
Comment (hoangdat) [reply mailbox/memory/src/main/resources/META-INF/spring/mailbox-memory.xml]: Duc will spend a litte time to explain this modification for me? It is related to `Spring`?

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 20]
```diff
@@ -1401,10 +1407,47 @@ void expungeShouldCallAllPreDeletionHooksOnEachMessageDeletionCall() throws Exce
                     .containsOnly(composeId1.getMessageId(), composeId2.getMessageId());
             }
 
+            @Test
+            void expungeShouldCallAllPreDeletionHooksOnlyOnMessagesMarkedAsDeleted() throws Exception {
+                ComposedMessageId composeId1 = inboxManager.appendMessage(AppendCommand.builder()
```
Comment (aduprat) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Can you add a message without DELETED?

Comment (chibenwa) [reply mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Yes, and I did it in an other test :trollface:

Comment (aduprat) [reply mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: I asked that because it's what the test name means, IMO.

Comment (chibenwa) [reply mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Yep looks like a intellij warning that got solved and I removed the all append... sorry and good catch..

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 21]
```diff
@@ -1300,4 +1333,190 @@ void closingSessionShouldWork() throws Exception {
...
+                    .extracting(deleteMetadata -> deleteMetadata.getMessageMetaData().getMessageId())
+                    .containsOnly(composeId1.getMessageId(), composeId2.getMessageId());
+            }
+
+            @Test
+            void expungeShouldCallAllPreDeletionHooksOnlyOnMessagesMarkedAsDeleted() throws Exception {
+                ComposedMessageId composeId1 = inboxManager.appendMessage(AppendCommand.builder()
+                    .withFlags(new Flags(Flags.Flag.DELETED))
+                    .build(message), session);
+                ComposedMessageId composeId2 = inboxManager.appendMessage(AppendCommand.builder()
```
Comment (aduprat) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Then you have a warning, _composeId2_ is not used, IMO ;)

**Comment by aduprat** [comment]
Was :green_apple: before CI restart.

**Comment by chibenwa** [comment]
Merged

File: mailbox/api/src/main/java/org/apache/james/mailbox/extension/PreDeletionHook.java [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+        private DeleteOperation(DeletionId deletionId, List<MetadataWithMailboxId> deletionMetadataList) {
+            this.deletionId = deletionId;
+            this.deletionMetadataList = deletionMetadataList;
+        }
+
+        public DeletionId getDeletionId() {
+            return deletionId;
+        }
+
+        public List<MetadataWithMailboxId> getDeletionMetadataList() {
```
Comment (hoangdat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/extension/PreDeletionHook.java]: @chibenwa @trantienduchn 
Do you think we need to refactor this this POJO to implementing a hook to store DeletedMessage into Vault?
IMO, with Metadata We lack many infomations to store DeletedMessage (e.g: Sender, Recipients, Attachment, ....)

Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/extension/PreDeletionHook.java]: Many of these information can be parsed from the mimeMessage - or from the mailbox context.

IMO you should have enough to be doing what you need to do.

Comment (hoangdat) [reply mailbox/api/src/main/java/org/apache/james/mailbox/extension/PreDeletionHook.java]: thanks
